pkg/restore/merge.go

```diff
@@ -0,0 +1,164 @@
+func MergeFileRanges(
+	files []*kvproto.File, splitSizeBytes, splitKeyCount uint64,
+) ([]rtree.Range, *MergeRangesStat, error) {
+	if len(files) == 0 {
+		return []rtree.Range{}, &MergeRangesStat{}, nil
+	}
+	totalBytes := uint64(0)
+	totalKvs := uint64(0)
+	totalFiles := len(files)
+	writeCFFile := 0
+	defaultCFFile := 0
+
+	filesMap := make(map[string][]*kvproto.File)
+	for _, file := range files {
+		filesMap[string(file.StartKey)] = append(filesMap[string(file.StartKey)], file)
+
+		// We skips all default cf files because we don't range overlap.
+		if file.Cf == writeCFName || strings.Contains(file.GetName(), writeCFName) {
+			writeCFFile++
+		} else if file.Cf == defaultCFName || strings.Contains(file.GetName(), defaultCFName) {
+			defaultCFFile++
+		}
+		totalBytes += file.TotalBytes
+		totalKvs += file.TotalKvs
+	}
+	if writeCFFile == 0 && defaultCFFile == 0 {
+		return []rtree.Range{}, nil, errors.Annotatef(berrors.ErrRestoreInvalidBackup,
+			"unknown backup data from neither Wrtie CF nor Default CF")
+	}
+
+	// RawKV does not have data in write CF.
+	totalRegions := writeCFFile
+	if defaultCFFile > writeCFFile {
+		totalRegions = defaultCFFile
+	}
+
+	// Check if files are overlapped
+	rangeTree := rtree.NewRangeTree()
+	for key := range filesMap {
+		files := filesMap[key]
+		if out := rangeTree.InsertRange(rtree.Range{
+			StartKey: files[0].GetStartKey(),
+			EndKey:   files[0].GetEndKey(),
+			Files:    files,
+		}); out != nil {
+			return nil, nil, errors.Annotatef(berrors.ErrRestoreInvalidRange,
+				"duplicate range %s files %+v", out, files)
+		}
+	}
+
+	needMerge := func(left, right *rtree.Range) bool {
+		leftBytes, leftKeys := left.BytesAndKeys()
+		rightBytes, rightKeys := right.BytesAndKeys()
+		if rightBytes == 0 {
+			return true
+		}
+		if leftBytes+rightBytes > splitSizeBytes {
+			return false
+		}
+		if leftKeys+rightKeys > splitKeyCount {
+			return false
+		}
+		// Do not merge ranges in different tables.
+		if tablecodec.DecodeTableID(kv.Key(left.StartKey)) != tablecodec.DecodeTableID(kv.Key(right.StartKey)) {
+			return false
+		}
+		// Do not merge ranges in different indexes even if they are in the same
+		// table, as rewrite rule only supports rewriting one pattern.
+		// tableID, indexID, indexValues, err
+		tableID1, indexID1, _, err1 := tablecodec.DecodeIndexKey(kv.Key(left.StartKey))
+		tableID2, indexID2, _, err2 := tablecodec.DecodeIndexKey(kv.Key(right.StartKey))
+		switch {
+		case (err1 != nil && err2 == nil) || (err1 == nil && err2 != nil):
+			return false
+		case err1 == nil && err2 == nil:
+			// Both of them are index keys.
+			if tableID1 != tableID2 || indexID1 != indexID2 {
+				// Do not merge if they are not the same index.
+				return false
+			}
+			// They are in the same index, merge left and right.
+			return true
+		default:
+			// err1 != nil && err2 != nil
+			// They are record keys, merge left and right.
+			return true
+		}
```

Comment:
```suggestion
		// If both of them are index keys, ...
		if err1 == nil && err2 == nil {
			// Merge left and right if they are in the same index.
			return indexID1 == indexID2
		}
		// Otherwise, merge if they are both record keys
		return err1 != nil && err2 != nil
```

we have also already checked the table ID in line 111 so `tableID1 != tableID2` is impossible.